### DISCUSSION TITLE
duckdb-wasm: ok under `pages/`, broken under `app/`

### DISCUSSION THREAD

**Discussion by ryan-williams** [vercel/next.js#57819] [App Router]
[duckdb-wasm](https://github.com/duckdb/duckdb-wasm) can load a Parquet file on the server in a `pages/` page ([demo](https://runsascoded.com/next-duckdb-parquet-demo)), but not in a similar `app/` page.

Full repro/details at [ryan-williams/next-duckdb-parquet-demo](https://github.com/ryan-williams/next-duckdb-parquet-demo#broken):

> ## ⚠️ Broken under "App Router" ⚠️ <a id="broken"></a>
> 
> For some reason, [the server-side `loadParquet` path][loadParquet] doesn't seem to work under Next.js's new "App router."
> 
> Here's a trimmed down example (from [the @app-broken branch][@app-broken]):
> 
> - [pages/pages.tsx] ✅ works
> - [app/app/page.tsx] ❌ broken
> 
> Attempting to load the latter emits this on the server side:
> ```
> duckdb-wasm fetch
> bestBundle: {
>   mainModule: '/Users/ryan/c/next-duckdb-parquet-demo/node_modules/@duckdb/duckdb-wasm/dist/duckdb-eh.wasm',
>   mainWorker: '/Users/ryan/c/next-duckdb-parquet-demo/node_modules/@duckdb/duckdb-wasm/dist/duckdb-node-eh.worker.cjs',
>   pthreadWorker: null
> }
> made logger
> made db
> worker terminated with 1 pending requests
> ```
> and the client hangs.
> 
> Similarly, `next build` times out, meaning I can't deploy a demo. [Here's a GitHub Actions run][failed GHA] that failed due to this issue.
> 
> `diff pages/pages.tsx app/app/page.tsx`:
> ```diff
> -export async function getStaticProps() {
> -    const data = await loadParquet<Person>(parquetPath)  // ✅ works fine under pages/
> -    return { props: { data } }
> -}
> -
> -export default function Home({ data }: { data: Person[] }) {
> +export default async function Home() {
> +    const data = await loadParquet<Person>(parquetPath)  // ❌ broken under app/
> ```
> 
> [This line][broken db instantiation line]:
> 
> ```typescript
> await db.instantiate(bundle.mainModule, bundle.pthreadWorker);
> ```
> emits this error:
> ```
> worker terminated with 1 pending requests
> ```
> 
> I haven't found any further leads about what might be going on.
> 
> 
> [runsascoded.com/next-duckdb-parquet-demo]: https://runsascoded.com/next-duckdb-parquet-demo
> [@app-broken]: https://github.com/ryan-williams/next-duckdb-parquet-demo/tree/app-broken
> [pages/pages.tsx]: https://github.com/ryan-williams/next-duckdb-parquet-demo/blob/app-broken/pages/pages.tsx
> [app/app/page.tsx]: https://github.com/ryan-williams/next-duckdb-parquet-demo/blob/app-broken/app/app/page.tsx
> [failed GHA]: https://github.com/ryan-williams/next-duckdb-parquet-demo/actions/runs/6708443526/job/18229259080
> [broken db instantiation line]: https://github.com/ryan-williams/next-duckdb-parquet-demo/blob/app-broken/src/parquet.ts#L43
> [loadParquet]: https://github.com/ryan-williams/next-duckdb-parquet-demo/blob/app-broken/src/parquet.ts#L66-L72
>

**Comment by ryan-williams** [comment]
I documented some more instances of the `worker terminated with 1 pending requests` issue at https://github.com/duckdb/duckdb-wasm/issues/1588.

**Reply by michaelwallabi** [reply]
Ran into this issue today as well within an `/api` route, always hangs on `instantiate` with the same messages you post.

**Reply by michaelwallabi** [reply]
Similar to @ryan-williams, I got here after discovering that using duckdb-node doesn't seem to work in Vercel due to lambda issues.

https://github.com/duckdb/duckdb-node/issues/15
